File: minikube/firmware/coordinate_loader.py
# Coordinate Loader - Parse and normalize LED positions from solution1.txt
# Format: LED_0004 0.170590 -1.000000 -0.077250

import config

def load_coordinates(filename):
    """
    Load and normalize LED coordinates from file.

    Args:
        filename: Path to coordinates file (solution1.txt)

    Returns:
        List of (x, y, z) tuples normalized to [0.0, 1.0], or None for missing IDs
        Index in list corresponds to LED ID from file
    """
    raw_coords = []

    print(f"Loading coordinates from {filename}...")

    # Pass 1: Parse all coordinates and find max LED ID
    try:
        with open(filename, 'r') as f:
            for line in f:
                line = line.strip()

                # Skip empty lines and comments
                if not line or line.startswith('#'):
                    continue

                # Only process LED lines
                if not line.startswith('LED_'):
                    continue

                parts = line.split()
                if len(parts) < 4:
                    continue

                # Extract LED ID from LED_0004 -> 4
                led_id_str = parts[0].split('_')[-1]
                led_id = int(led_id_str)

                # Parse coordinates
                x = float(parts[1])
                y = float(parts[2])
                z = float(parts[3])

                raw_coords.append((led_id, x, y, z))

    except OSError as e:
        print(f"Error loading coordinates: {e}")
        return []

    if not raw_coords:
        print("No LED coordinates found in file!")
        return []

    print(f"Loaded {len(raw_coords)} LED positions")

    # Find max LED ID to size the coordinate array
    max_led_id = max(led_id for led_id, _, _, _ in raw_coords)
    print(f"Max LED ID in file: {max_led_id}")

    # Initialize coordinate list sized for all LEDs in file
    coords = [None] * (max_led_id + 1)

    # Pass 2: Find bounding box
    xs = [c[1] for c in raw_coords]
    ys = [c[2] for c in raw_coords]
    zs = [c[3] for c in raw_coords]

    x_min, x_max = min(xs), max(xs)
    y_min, y_max = min(ys), max(ys)
    z_min, z_max = min(zs), max(zs)

    print(f"Bounding box:")
    print(f"  X: [{x_min:.3f}, {x_max:.3f}]")
    print(f"  Y: [{y_min:.3f}, {y_max:.3f}]")
    print(f"  Z: [{z_min:.3f}, {z_max:.3f}]")

    # Calculate ranges
    x_range = x_max - x_min
    y_range = y_max - y_min
    z_range = z_max - z_min

    # Pass 3: Normalize and store
    for led_id, x, y, z in raw_coords:
        # Normalize to [0.0, 1.0]
        # Handle zero range (all coordinates same) by defaulting to 0.5
        nx = (x - x_min) / x_range if x_range > 0 else 0.5
        ny = (y - y_min) / y_range if y_range > 0 else 0.5
        nz = (z - z_min) / z_range if z_range > 0 else 0.5

        coords[led_id] = (nx, ny, nz)

    # Count how many LEDs have coordinates
    loaded_count = sum(1 for c in coords if c is not None)
    print(f"Normalized {loaded_count} LED coordinates to [0.0, 1.0] range")

    return coords


def get_coordinate(coords, led_id):
    """
    Get normalized coordinate for a specific LED.

    Args:
        coords: List from load_coordinates()
        led_id: LED index (0-73)

    Returns:
        (x, y, z) tuple or None if LED has no coordinate
    """
    if led_id < 0 or led_id >= len(coords):
        return None
    return coords[led_id]

File: showrunner/lib/coordinate-loader.js
import { readFile } from 'fs/promises';

/**
 * LED Coordinate Loader
 * Parses LED coordinate files and normalizes coordinates to [0, 1] range
 */
export class CoordinateLoader {
  constructor(filePath) {
    this.filePath = filePath;
    this.coordinates = []; // Sparse array: ledId -> [x, y, z] or undefined
    this.rawCoordinates = []; // [{id, x, y, z}]
    this.ledCount = 0;
  }

  /**
   * Load and parse LED coordinate file
   */
  async load() {
    const content = await readFile(this.filePath, 'utf-8');
    const lines = content.split('\n');

    const rawCoords = [];

    for (const line of lines) {
      const trimmed = line.trim();

      // Skip empty lines and comments
      if (!trimmed || trimmed.startsWith('#')) {
        continue;
      }

      // Only process lines starting with LED_
      if (!trimmed.startsWith('LED_')) {
        continue;
      }

      // Parse line: LED_<channel>_<id> x y z or LED_<id> x y z
      const parts = trimmed.split(/\s+/);
      if (parts.length < 4) {
        console.warn(`Skipping invalid line: ${line}`);
        continue;
      }

      const ledIdPart = parts[0];
      const x = parseFloat(parts[1]);
      const y = parseFloat(parts[2]);
      const z = parseFloat(parts[3]);

      // Extract LED ID from last underscore
      const lastUnderscoreIndex = ledIdPart.lastIndexOf('_');
      if (lastUnderscoreIndex === -1) {
        console.warn(`Invalid LED ID format: ${ledIdPart}`);
        continue;
      }

      const ledId = parseInt(ledIdPart.substring(lastUnderscoreIndex + 1), 10);

      if (isNaN(ledId) || isNaN(x) || isNaN(y) || isNaN(z)) {
        console.warn(`Invalid values in line: ${line}`);
        continue;
      }

      rawCoords.push({ id: ledId, x, y, z });
    }

    if (rawCoords.length === 0) {
      throw new Error('No valid LED coordinates found in file');
    }

    this.rawCoordinates = rawCoords;
    this.normalizeCoordinates();
    this.ledCount = rawCoords.length;

    console.log(`Loaded ${this.ledCount} LED coordinates from ${this.filePath}`);
  }

  /**
   * Normalize coordinates to [0.0, 1.0] range
   * Formula: normalized = (value - min) / (max - min)
   */
  normalizeCoordinates() {
    if (this.rawCoordinates.length === 0) {
      return;
    }

    // Find bounding box
    let xMin = Infinity, xMax = -Infinity;
    let yMin = Infinity, yMax = -Infinity;
    let zMin = Infinity, zMax = -Infinity;

    for (const coord of this.rawCoordinates) {
      xMin = Math.min(xMin, coord.x);
      xMax = Math.max(xMax, coord.x);
      yMin = Math.min(yMin, coord.y);
      yMax = Math.max(yMax, coord.y);
      zMin = Math.min(zMin, coord.z);
      zMax = Math.max(zMax, coord.z);
    }

    const xRange = xMax - xMin;
    const yRange = yMax - yMin;
    const zRange = zMax - zMin;

    // Normalize and store in sparse array
    for (const coord of this.rawCoordinates) {
      const normalizedX = xRange > 0 ? (coord.x - xMin) / xRange : 0.5;
      const normalizedY = yRange > 0 ? (coord.y - yMin) / yRange : 0.5;
      const normalizedZ = zRange > 0 ? (coord.z - zMin) / zRange : 0.5;

      this.coordinates[coord.id] = [normalizedX, normalizedY, normalizedZ];
    }

    console.log(`Normalized coordinates to [0, 1] range`);
    console.log(`  X range: [${xMin.toFixed(3)}, ${xMax.toFixed(3)}]`);
    console.log(`  Y range: [${yMin.toFixed(3)}, ${yMax.toFixed(3)}]`);
    console.log(`  Z range: [${zMin.toFixed(3)}, ${zMax.toFixed(3)}]`);
  }

  /**
   * Get normalized coordinate for a specific LED ID
   * @param {number} ledId - LED index (0 to N-1)
   * @returns {Array|null} [x, y, z] normalized coordinates or null if missing
   */
  getCoordinate(ledId) {
    return this.coordinates[ledId] || null;
  }

  /**
   * Get total number of LEDs with valid coordinates
   * @returns {number} LED count
   */
  getLEDCount() {
    return this.ledCount;
  }

  /**
   * Check if a LED has coordinates
   * @param {number} ledId - LED index
   * @returns {boolean} True if LED has coordinates
   */
  hasCoordinate(ledId) {
    return this.coordinates[ledId] !== undefined;
  }

  /**
   * Get all LED IDs that have coordinates
   * @returns {Array} Array of LED IDs
   */
  getAllLEDIds() {
    const ids = [];
    for (let i = 0; i < this.coordinates.length; i++) {
      if (this.coordinates[i] !== undefined) {
        ids.push(i);
      }
    }
    return ids;
  }
}
